README.md

 ![Platform](https://img.shields.io/badge/platform-macos%20|%20linux-green.svg)
 ![Language](https://img.shields.io/badge/language-c++-red.svg)
 ![Libraries](https://img.shields.io/badge/libraries-cad%20|%20geant4-blue.svg)



<b>GE</b>ant4 <b>M</b>onte-<b>C</b>arlo


<b>gemc</b> is an application based on <a href="https://geant4.cern.ch"> Geant4 </a> to simulate the 
passage of particles through matter.

It provides:

* application independent geometry description
* easy interface to build / run experiments
* cad/gdml imports


<p align="center">
	<img src="https://github.com/gemc/docs/blob/master/webPage/source/beam.png"   height="170" width="170">
	<img src="https://github.com/gemc/docs/blob/master/webPage/source/eic_beam.png" height="170" width="170">
	<img src="https://github.com/gemc/docs/blob/master/webPage/source/eic.png"    height="170" width="170">
	<img src="https://github.com/gemc/docs/blob/master/webPage/source/bubble.png" height="170" width="170">
</p>
<i> examples of experiments using gemc. From left to right: a <a href="https://www.jlab.org/Hall-B/clas12-web/">CLAS12</a> beamline event; 
The electron and ion beamline and interation point detectors for the <a href="https://www.jlab.org/meic/">electron-ion collider</a>.
The gemc simulation of the Electron Ion Collider beamline and detectors.
10,000 electrons producing photons in the 6mm collimator in the bubble experiments at Jefferson Lab.</i>

========


### Overview:

gemc makes easy things trivial and hard things possible.

Users can build and run complex setups with minimal programming knowledge. 
See for example <a href="https://gemc.jlab.org/gemc/html/examples/paddles.html#simplepaddleexample">how to build a TOF with
few lines of code</a>.

Experiments can be loaded  using a combination of several available factories:

- MYSQL
- TEXT
- GDML
- CAD (STL, PLY, OBJ formats)
- C++ Plugin

<p align="center">
	<a href="https://github.com/gemc/detectors/blob/master/humanBody/Upper_GI.stl">
	<img src="https://github.com/gemc/docs/blob/master/webPage/source/examples/humanBody.png" width="380px" height="380px"></img></a>
	<a href="https://github.com/gemc/detectors/blob/master/forFun/enterprise.stl"> 
	<img src="https://github.com/gemc/docs/blob/master/webPage/source/examples/forFun.png"    width="380px" height="380px"></img></a><br>
	<small><i> gemc can <a href="https://gemc.jlab.org/gemc/html/documentation/gdmlCadFactories.html">import models from CAD and GDML</a>.
   	Left: the upper gastrointestinal system is modeled in CAD.
   	It can be <a href="https://gemc.jlab.org/gemc/html/examples/humanBody.html">imported in GEMC and made it sensitive</a> 
		so that radiation doses can be measured.
   	Right: the mighty USS Enterprise NCC 1701-A <a href="https://gemc.jlab.org/gemc/html/examples/forFun.html">can be
   	used to shoot protons torpedos</a>. 
   </i></small>
</p>

### Simulations are application independent

Once the user defined setup is loaded, it is translated in geant4. This includes:

- geometry
- materials
- mirrors
- physics list
- digitization
- electromagnetic fields

All particles are transported through matters and
produce radiation, hits, secondaries. The geant4 results are then collected and organized according to user preferences.



<p align="center">
	<img src="https://github.com/gemc/docs/blob/master/webPage/source/gemcArchitecture.png" width="90%">
</p>




### Documentation:
* <a href="http://gemc.jlab.org"> gemc home page </a>





changeLog
Changelog for GEMC
==================


4/7/2021

 - phi-segmented field routine fix
 - hipo event buffer increased

4/1/2021

 - writing true information in hipo format

3/31/2021

 - renaming utils and options filenames to gemcUtils and gemcOption

3/29/2021

 - introducing exit codes for gemc


3/25/2021

 - removing unecessary logs on stepping action. return 0 not 1 from gemc.cc


3/24/2021

 - correct conversion factor to time for ctof and cnd
 - SKIPNGEN now completely skips all events before its value, previsouly header and generated particles were written

3/23/2021

 - Make CHECK_OVERLAPS=1 work again.


3/18/2021

 - addressing issue 190, gemc will quit if argument is unknown / not found
 - addressing repeating log
 - addressing issue 191, density of vacuumOpt


3/17/2021

 - strict check on scan_number


3/16/2021

 - exit if pid in input file is wrong

2/16/2021

 - solenoid and torus field in hipo RUN::config

2/5/2021

 - addressing gcc warnings


1/28/2021

 - ec and pcal hit process are merged


12/8/2020

 - CLAS12 detectors hipo output


10/13/2020

 - starting hipo4 output: definind schemas


10/6/2020

 - writing RUN::config bank


8/31/2020

- ALERT digitization improvements


8/7/2020

- added muMinusCaptureAtRest


7/29/2020

- detail on error for each coordinate. fix typo on coordinate 3. add tesla units on log


7/28/2020

 - HARDWARESTATUS implementation for CLAS12 detectors

7/13/2020

 - fixed ROTATE_FIELDMAP option

7/10/2020

 - added RECORD_MIRRORS to control mirror type recording

7/7/2020

 - added hydrogen gas material

7/6/2020

 - added options.html to repository


7/2/2020

 - added alert hit processes
 - improving CLAS12 dc smearing and inefficiency


.. Note::
   6/23/20 gemc 2.8 released.

|

6/16/2020

 - added tritium isotope


6/03/2020

 - added DIGITIZATION_TIMESTAMP option to pass to digitization routines


5/19/2020

 - fix cylindrical-x and cylindrical-y field map bug


5/14/2020

 - CLAS12 DC hit process updates: b-field correction to local angle, drift-velocity from t2d table
 - CLAS12 Implement status tables and revert to GSIM photoelectron variance generator

4/20/2020

 - added ALERT atof hit process


3/31/2020

 - CLAS12 added global offset for ec, pcal and htcc


3/27/2020

 - CLAS12 using non hardcoded run number and digiVariation all detectors


2/25/2020

 - CLAS12 dc T0corrections


2/10/2020

- added  SHIFT_LUND_VERTEX option to shift a file generator vertex.

  example: -SHIFT_LUND_VERTEX="(0, 0, -3)cm"


2/4/2020

 - added entry "doubleRotation" for volume rotation: this must be followed
   by a set of 6 numbers with units, describing two consecutive euler rotations.

1/9/2020

 - GUI background color changed to white
 - Neutrals particles color changed to blue
 - FTOF resolution coming from CCDB
 - Double radius for hits above thresholds
 - added hBertiniCaptureAtRest to process catalogueq

12/18/19

 - Modified RTPC hit process parameters and bank names.


10/7/19

 - added hBertiniCaptureAtRest tag

7/23/19

 - exiting if gcard is not found

7/10/19

 - added DIGITIZATION_VARIATION, to be used by digitization routines.

7/8/19

 - added background merging for FTOF, EC, PCAL, FT_CAL, FT_HODO, LTCC, HTCC

6/17/19

 - number of events made long int

6/14/19

- added target position from CCDB (fixed variation, todo: detector-dependent variation mechanism) to MM and CTOF routines.

5/23/19

 - added "PhotoNuclear" entry in the physics option list as a check that this part is not included with the normal phys list constructors.

5/17/19

 - option RECORD_MIRRORS renamed RECORD_OPTICALPHOTONS

5/8/19

 - added JSON ouput for all generator options

5/7/19

 - first implementation of JSON output for the options
 - output implemention of JSON options for txt and evio

5/6/19

 - reorg of gemc_option categories
 - only write option map out if output is specified
 - adding json format output for option map (in process)

5/1/19

 - fixed dc setting of NWIRES.
 - added flight time to tdc in dc


4/25/19

 - Richard Holmes: added FILTER_HADRONS options: when set to 1, do not write events if there are no hadrons.
   Otherwise if nonzero write only events having a hadron with matching ID. For example -FILTER_HADRONS=2212 for protons.
   Also, do not print progress if evtN hasn't changed.

4/18/19

 - added mirror hit process routine as standalone routine and bank

3/26/19

 - added FILTER_HITS options: when set to 1, do not write events if there are not hits.

3/25/19

- making sure initWithRunNumber is not loading calibration constants before activating event for all hit processes

2/25/19

 - Updated LUND documentation on Info tab

1/31/19

 - Richard Holmes: The RERUN_SELECTED option allows one to rerun multiple events whose RNG state was saved in a previous pass.

    For example, suppose there is a rare background signature of an electron between 100 MeV and 1 GeV in detector ID 1000
    and you want to visualize a few dozen such events to investigate where the background is coming from. You can do a batch
    run using


output/gbank.cc
/// \file gbank.cc
/// Contains:
/// - read_banks: reads the Hit Process Map and
/// builds the gemc bank class map.\n\n
/// \author \n Maurizio Ungaro
/// \author mail: ungaro@jlab.org\n\n\n

// Qt headers
#include <QtSql>

// gemc headers
#include "gbank.h"
#include "string_utilities.h"
#include "gemcUtils.h"

// Variable Type is two chars.
// The first char:
//  R for raw integrated
//  D for dgt integrated
//  S for raw step by step
//  M for digitized multi-hit
//  C for charge, time
//  N for not relevant

// The second char:
// i for integers
// d for doubles
// s for strings


map<string, gBank> read_banks(goptions gemcOpt, map<string, string> allSystems)
{
	double verbosity     = gemcOpt.optMap["BANK_VERBOSITY"].arg ;
	
	// geant4 information, integrated over
	map<string, gBank> banks;
	
	gBank abank;
	
	// event header
	abank = gBank(HEADER_BANK_TAG, "header", "Data Header Bank");
	abank.load_variable("time",       1, "Ns", "Time-stamp");
	abank.load_variable("runNo",      2, "Ni", "Run Number");
	abank.load_variable("evn",        3, "Ni", "Event Number");
	abank.load_variable("evn_type",   4, "Ni", "Event Type. 1 for physics events, 10 for scaler. Negative sign for MC.");
	abank.load_variable("beamPol",    5, "Nd", "Beam Polarization");
	abank.orderNames();
	banks["header"] = abank;
	
	// event header
	abank = gBank(USER_HEADER_BANK_TAG, "userHeader", "User Header Bank");
	abank.orderNames();
	banks["userHeader"] = abank;
	
	// generated particle infos
	abank =  gBank(GENERATED_PARTICLES_BANK_TAG, "generated", "Generated Particles");
	abank.load_variable("pid",           1,  "Ni", "Particle ID");
	abank.load_variable("px",            2,  "Nd", "x component of momentum");
	abank.load_variable("py",            3,  "Nd", "y component of momentum");
	abank.load_variable("pz",            4,  "Nd", "z component of momentum");
	abank.load_variable("vx",            5,  "Nd", "x component of vertex");
	abank.load_variable("vy",            6,  "Nd", "y component of vertex");
	abank.load_variable("vz",            7,  "Nd", "z component of vertex");
	abank.load_variable("time",          8,  "Nd", "time");
	abank.load_variable("multiplicity",  9,  "Ni", "number of particles at this vertex");
	abank.orderNames();
	banks["generated"] = abank;
	
	// RF info
	abank =  gBank(RF_BANK_TAG, "rf", "RF Signals");
	abank.load_variable("id",    1,  "Ni", "RF ID");
	abank.load_variable("rf",    2,  "ND", "RF ID");
	banks["rf"] = abank;
	
	// particle summary infos
	// this is a daughter bank of the generated particle infos
	abank =  gBank(GENERATED_SUMMARY_BANK_TAG, "psummary", "Generated Particles Summary");
	abank.load_variable("dname",    1, "Ns", "Detector Name");
	abank.load_variable("stat",     2, "Ni", "Number of Hits in the detector");
	abank.load_variable("etot",     3, "Nd", "Total Energy Deposited");
	abank.load_variable("t",        4, "Nd", "Fastest Time on Detector");
	abank.load_variable("nphe",     5, "Ni", "Number of Photoelectrons");
	abank.load_variable("upx",     10, "Nd", "unsmeared px");
	abank.load_variable("upy",     11, "Nd", "unsmeared py");
	abank.load_variable("upz",     12, "Nd", "unsmeared pz");
	abank.load_variable("spx",     13, "Nd", "smeared px");
	abank.load_variable("spy",     14, "Nd", "smeared py");
	abank.load_variable("spz",     15, "Nd", "smeared pz");
	abank.orderNames();
	banks["psummary"] = abank;
	
	
	// all hit bank information have a variable with num 0 "hitn" = hit number;
	
	// geant4 raw integrated
	// common for all banks
	abank =  gBank(RAWINT_ID, "raws", "Geant4 true information integrated over the hit");
	abank.load_variable("pid",     1,   "Ri", "ID of the first particle entering the sensitive volume");
	abank.load_variable("mpid",    2,   "Ri", "ID of the mother of the first particle entering the sensitive volume");
	abank.load_variable("tid",     3,   "Ri", "Track ID of the first particle entering the sensitive volume");
	abank.load_variable("mtid",    4,   "Ri", "Track ID of the mother of the first particle entering the sensitive volume");
	abank.load_variable("otid",    5,   "Ri", "Track ID of the original track that generated the first particle entering the sensitive volume");
	abank.load_variable("trackE",  6,   "Rd", "Energy of the track");
	abank.load_variable("totEdep", 7,   "Rd", "Total Energy Deposited (in MeV");
	abank.load_variable("avg_x",   8,   "Rd", "Average X position in the global reference system (in mm)");
	abank.load_variable("avg_y",   9,   "Rd", "Average Y position in the global reference system (in mm)");
	abank.load_variable("avg_z",  10,   "Rd", "Average Z position in the global reference system (in mm)");
	abank.load_variable("avg_lx", 11,   "Rd", "Average X position in the local reference system (in mm)");
	abank.load_variable("avg_ly", 12,   "Rd", "Average Y position in the local reference system (in mm)");
	abank.load_variable("avg_lz", 13,   "Rd", "Average Z position in the local reference system (in mm)");
	abank.load_variable("px",     14,   "Rd", "x component of momentum of the particle entering the sensitive volume");
	abank.load_variable("py",     15,   "Rd", "y component of momentum of the particle entering the sensitive volume");
	abank.load_variable("pz",     16,   "Rd", "z component of momentum of the particle entering the sensitive volume");
	abank.load_variable("vx",     17,   "Rd", "x component of point of origin of the particle entering the sensitive volume");
	abank.load_variable("vy",     18,   "Rd", "y component of point of origin of the particle entering the sensitive volume");
	abank.load_variable("vz",     19,   "Rd", "z component of point of origin of the particle entering the sensitive volume");
	abank.load_variable("mvx",    20,   "Rd", "x component of point of origin the mother of the particle entering the sensitive volume");
	abank.load_variable("mvy",    21,   "Rd", "y component of point of origin of the mother of the particle entering the sensitive volume");
	abank.load_variable("mvz",    22,   "Rd", "z component of point of origin of the mother of the particle entering the sensitive volume");
	abank.load_variable("avg_t",  23,   "Rd", "Average time");
	abank.load_variable("nsteps", 24,   "Ri", "Number of geant4 steps");
	abank.load_variable("procID", 25,   "Ri", "Process that created the particle. It's an integer described at gemc.jlab.org");
	abank.load_variable("hitn",   99,   "Ri", "Hit Number");
	abank.orderNames();
	banks["raws"] = abank;
	
	
	// geant4 raw step by step
	// common for all banks
	abank =  gBank(RAWSTEP_ID, "allraws", "Geant4 true information step by step");
	abank.load_variable("pid",     1,   "Si", "ID of the particle in the sensitive volume");
	abank.load_variable("mpid",    2,   "Si", "ID of the mother in the sensitive volume");
	abank.load_variable("tid",     3,   "Si", "Track ID of the particle in the sensitive volume");
	abank.load_variable("mtid",    4,   "Si", "Track ID of the mother of the particle in the sensitive volume");
	abank.load_variable("otid",    5,   "Si", "Track ID of the original track that generated the particle in the sensitive volume");
	abank.load_variable("trackE",  6,   "Sd", "Energy of the track");
	abank.load_variable("edep",    7,   "Sd", "Energy Deposited");
	abank.load_variable("x",       8,   "Sd", "X position in global reference system");
	abank.load_variable("y",       9,   "Sd", "Y position in global reference system");
	abank.load_variable("z",      10,   "Sd", "Z position in global reference system");
	abank.load_variable("lx",     11,   "Sd", "X position in local reference system");
	abank.load_variable("ly",     12,   "Sd", "Y position in local reference system");
	abank.load_variable("lz",     13,   "Sd", "Z position in local reference system");
	abank.load_variable("px",     14,   "Sd", "x component of momentum of the particle in the sensitive volume");
	abank.load_variable("py",     15,   "Sd", "y component of momentum of the particle in the sensitive volume");
	abank.load_variable("pz",     16,   "Sd", "z component of momentum of the particle in the sensitive volume");
	abank.load_variable("vx",     17,   "Sd", "x component of primary vertex of the particle in the sensitive volume");
	abank.load_variable("vy",     18,   "Sd", "y component of primary vertex of the particle in the sensitive volume");
	abank.load_variable("vz",     19,   "Sd", "z component of primary vertex of the particle in the sensitive volume");
	abank.load_variable("mvx",    20,   "Sd", "x component of primary vertex of the mother of the particle in the sensitive volume");
	abank.load_variable("mvy",    21,   "Sd", "y component of primary vertex of the mother of the particle in the sensitive volume");
	abank.load_variable("mvz",    22,   "Sd", "z component of primary vertex of the mother of the particle in the sensitive volume");
	abank.load_variable("t",      23,   "Sd", "time");
	abank.load_variable("stepn",  98,   "Si", "step index");
	abank.load_variable("hitn",   99,   "Si", "Hit Number");
	abank.orderNames();
	banks["allraws"] = abank;
	
	// geant4 raw step by step
	// common for all banks
	abank =  gBank(CHARGE_TIME_ID, "chargeTime", "charge and time as seen by the electronics");
	abank.load_variable("id",     1,    "Ci", "hit identifier");
	abank.load_variable("q",      2,    "Cd", "charge as seen by electronics");
	abank.load_variable("t",      3,    "Cd", "time as seen by electronics");
	abank.load_variable("stepi",  98,   "Ci", "step index");
	abank.load_variable("hitn",   99,   "Ci", "Hit Number");
	banks["chargeTime"] = abank;
	
	
	// flux bank digitized infos
	// flux digitized provide just one "digitized" variable, the detector id
	abank =  gBank(FLUX_BANK_TAG, "flux", "Geant4 flux digitized information");
	abank.load_variable("hitn",   99,  "Di", "Hit Number");
	abank.load_variable("id",     1,  "Di", "ID of flux element");
	abank.orderNames();
	banks["flux"]   = abank;
	
	// mirror bank digitized infos
	// mirror digitized provide just one "digitized" variable, the detector id
	abank =  gBank(MIRROR_BANK_TAG, "mirror", "Geant4 mirror digitized information");
	abank.load_variable("hitn",   99,  "Di", "Hit Number");
	abank.load_variable("id",     1,  "Di", "ID of flux element");
	abank.orderNames();
	banks["mirror"] = abank;
	
	// counter bank integrated digitized infos
	// flux digitized provide just one "digitized" variable, the detector id
	abank =  gBank(COUNTER_BANK_TAG, "counter", "Geant4 counter digitized information");
	abank.load_variable("id",              1,  "Di", "ID of counter element");
	abank.load_variable("hitn",           99,  "Di", "Hit Number");
	abank.load_variable("ngamma",          10,  "Di", "number of gamma");
	abank.load_variable("nep",             11,  "Di", "number of electrons");
	abank.load_variable("nem",             12,  "Di", "number of positrons");
	abank.load_variable("npip",            13,  "Di", "number of pi+");
	abank.load_variable("npim",            14,  "Di", "number of pi-");
	abank.load_variable("npi0",            15,  "Di", "number of pi0");
	abank.load_variable("nkp",             16,  "Di", "number of k+");
	abank.load_variable("nkm",             17,  "Di", "number of k-");
	abank.load_variable("nk0",             18,  "Di", "number of k0");
	abank.load_variable("nproton",         19,  "Di", "number of protons");
	abank.load_variable("nneutron",        20,  "Di", "number of neutrons");
	abank.load_variable("nopticalphoton",  21,  "Di", "number of optical photons");
	abank.orderNames();
	banks["counter"]   = abank;
	
	
	// ancestors bank
	// Information about ancestral trajectories
	abank =  gBank (ANCESTORS_BANK_TAG, "ancestors", "Geant4 ancestors information");
	abank.load_variable ("pid",     1,   "Ri", "ID of the ancestor");
	abank.load_variable ("tid",     2,   "Ri", "Track ID of the ancestor");
	abank.load_variable ("mtid",    3,   "Ri", "Track ID of the mother of the ancestor");
	abank.load_variable ("trackE",  4,   "Rd", "Energy of the ancestor");
	abank.load_variable ("px",      5,   "Rd", "x component of momentum of the ancestor");
	abank.load_variable ("py",      6,   "Rd", "y component of momentum of the ancestor");
	abank.load_variable ("pz",      7,   "Rd", "z component of momentum of the ancestor");
	abank.load_variable ("vx",      5,   "Rd", "x component of vertex of the ancestor");
	abank.load_variable ("vy",      6,   "Rd", "y component of vertex of the ancestor");
	abank.load_variable ("vz",      7,   "Rd", "z component of vertex of the ancestor");
	abank.orderNames();
	banks["ancestors"]   = abank;
	
	
	// Loading all banks related to a system
	// then checking that all sensitive detectors have a bank
	for(map<string, string>::iterator sit = allSystems.begin(); sit != allSystems.end(); sit++)
	{
		string systemName    = sit->first;
		string systemFactory = sit->second;
		
		// these are already loaded
		if(systemName == "flux" || systemName == "mirror" || systemName == "counter") continue;
		
		// text factory
		if(systemFactory == "TEXT")
		{
			
			string fname = systemName + "__bank.txt";
			ifstream IN(fname.c_str());
			if(!IN)
			{
				// if file is not found, maybe it's in the GEMC_DATA_DIR directory
				if(getenv("GEMC_DATA_DIR")  != NULL)
				{
					fname = (string) getenv("GEMC_DATA_DIR") + "/" + fname;
					IN.open(fname.c_str());
				}
				
			}
			// now file should be loaded
			if(IN)
			{
				if(verbosity > 1)
					cout << "   > Loading bank TEXT definitions for <" << systemName << ">." << endl;
				
				
				// first get all banks for this system
				vector<string> banksForSystem;
				while(!IN.eof())
				{
					string dbline;
					getline(IN, dbline);
					
					if(!dbline.size()) continue;
					
					gtable gt(getStringVectorFromStringWithDelimiter(dbline, "|"));
					
					if(gt.data.size())
						if(gt.data[1] == "bankid")
							banksForSystem.push_back(gt.data[0]); // 0: bank name
				}
				// rewind IN
				IN.clear();
				IN.seekg(0);
				
				// now loading bank and variables
				for(unsigned b=0; b<banksForSystem.size(); b++)
				{
					while(!IN.eof())
					{
						string dbline;
						getline(IN, dbline);
						
						if(!dbline.size()) continue;
						
						gtable gt(getStringVectorFromStringWithDelimiter(dbline, "|"));
						
						// the bankid entry is always the first
						if(gt.data.size())
							if(gt.data[0] == banksForSystem[b])
							{
								string bname = gt.data[0];             // 0: bank name
								string vname = gt.data[1];             // 0: variable name
								string desc  = gt.data[2];             // 2: bank/variable description
								int num      = get_number(gt.data[3]); // 3: variable num is bank id
								string type  = gt.data[4];             // 4: variable type
								
								if(vname == "bankid")
								{
									abank =  gBank(num, bname, desc);
								}
								else
								{
									abank.load_variable(vname, num, type, desc);
									
								}
							}
					}
					abank.orderNames();
					banks[banksForSystem[b]] = abank;
					IN.clear();
					IN.seekg(0);
				}
				
				IN.close();
			} else {
				if(verbosity>2) {
					cout << "  !!! Error: Failed to open system bank file " << fname
					<< ". Maybe the filename doesn't exist? Exiting." << endl;
				}
			}
		}
		
		if(systemFactory == "MYSQL")
		{
			// connection to the DB
			QSqlDatabase db = openGdb(gemcOpt);
			string tname    = systemName + "__bank";
			// hardcoding original variation for now
			string variation = "original";
			
			if(verbosity > 1) cout << "   > Loading MYSQL definitions for <" << systemName << ">." << endl;
			
			// first getting all banks from system
			string dbexecute  = "select bankname, name from " + tname ;
			dbexecute += " where variation ='" + variation + "'";
			
			QSqlQuery q;
			if(!q.exec(dbexecute.c_str()))
			{
				cout  << " !!! Failed to execute MYSQL query " << dbexecute <<  ". This is a fatal error. Exiting." << endl;
				qDebug() << q.lastError();
				exit(0);
			}
			
			// Warning if nothing is found
			if(q.size() == 0 && verbosity)
			{
				cout << "  ** WARNING: system  \"" << systemName << "\" not found in MYSQL database "
				<< " for variation " << variation << endl << endl;
			}
			
			vector<string> banksForSystem;
			while (q.next())
			{
				// Reading variables
				
				// 0: variable name
				string name = qv_tostring(q.value(1));
				
				if(name == "bankid")
					banksForSystem.push_back(qv_tostring(q.value(0)));
				
			}
			
			
			for(unsigned b=0; b<banksForSystem.size(); b++)
			{
				// re-executing the query and loading variables
				dbexecute  = "select bankname, name, num, type, description from " + tname ;
				dbexecute += " where variation ='" + variation + "'";
				dbexecute += " and bankname = '" + banksForSystem[b]  + "'";
				q.exec(dbexecute.c_str());
				while (q.next())
				{
					string bname       = qv_tostring(q.value(0));  // 0: bank name
					string vname       = qv_tostring(q.value(1));  // 1: variable name
					int num = get_number(qv_tostring(q.value(2))); // 2: variable num is bank id
					string type =        qv_tostring(q.value(3));  // 3: variable type
					string desc =        qv_tostring(q.value(4));  // 4: variable description
					
					// need to make sure bankid is the first entry
					// that comes out of the mysql query
					if(vname == "bankid")
					{
						abank =  gBank(num, bname, desc);
					}
					else
					{
						abank.load_variable(vname, num, type, desc);
					}
				}
				abank.orderNames();
				banks[banksForSystem[b]] = abank;
				
			}
			
			
			// closing DB connection
			closeGdb(db);
		}
		
	}
	if(verbosity > 3)
	{
		for(map<string, gBank>::iterator it = banks.begin(); it != banks.end(); it++)
		cout << it->second;
	}
	
	return banks;
}

// Load a variable in the bank definition
void gBank::load_variable(string n, int i, string t, string d)
{
	// adding the variable index to order the map
	name.push_back(n);
	gid.push_back(i);
	type.push_back(t);
	description.push_back(d);
}


int gBank::getVarId(string bank) {
	for(unsigned int i=0; i<name.size(); i++)
	{
		if(name[i].find(bank) == 0) return gid[i];
	}
	return -1;
}



string gBank::getVarType(string var) {
	for(unsigned int i=0; i<name.size(); i++) {
		if(name[i] == var && type[i].length() == 2) {
			if(type[i].find("i") == 1) return "i";
			if(type[i].find("d") == 1) return "d";
		}
	}
	return "na";
}

int gBank::getVarBankType(string var) {
	for(unsigned int i=0; i<name.size(); i++) {
		if(name[i] == var  && type[i].length() == 2) {
			if(type[i].find("R") == 0) return RAWINT_ID;
			if(type[i].find("D") == 0) return DGTINT_ID;
			if(type[i].find("S") == 0) return RAWSTEP_ID;
			if(type[i].find("M") == 0) return DGTMULTI_ID;
			if(type[i].find("C") == 0) return CHARGE_TIME_ID;
		}
	}
	return 0;
}





// order names based on their ID
void gBank::orderNames()
{
	int minId = 1000;
	int maxId = 0;
	
	// first find min, max ID
	for(unsigned i=0; i<gid.size(); i++)
	{
		if(gid[i] < minId) minId = gid[i];
		if(gid[i] > maxId) maxId = gid[i];
	}
	
	int j = 0;
	for(int i=minId; i<=maxId; i++) {
		for(unsigned k=0; k<gid.size(); k++) {
			if(i == gid[k]) {
				orderedNames[j++] = name[k];
			}
		}
	}
}

// get bank definitions (all)
gBank getBankFromMap(string name, map<string, gBank>* banksMap) {
	if(banksMap->find(name) == banksMap->end()) {
		cout << "   !!! Error: >" << name << "< bank definitions not found. Exiting." << endl;
		exit(0);
	}
	
	return (*banksMap)[name];
}


// get dgt bank definitions
gBank getDgtBankFromMap(string name, map<string, gBank>* banksMap) {
	gBank thisBank, dgtBank;
	if(banksMap->find(name) == banksMap->end()) {
		cout << "   !!! Error: >" << name << "< dgt bank definitions not found. Exiting." << endl;
		exit(0);
	} else {
		// thisBank may have definitions other than DGT
		// so I'm extracting just the DGT variables from it.
		thisBank = (*banksMap)[name];
		
		dgtBank = gBank(DGTINT_ID, thisBank.bankName, thisBank.bdescription);
		for(unsigned int i=0; i<thisBank.name.size(); i++) {
			if(thisBank.getVarBankType(thisBank.name[i]) == DGTINT_ID) {
				dgtBank.load_variable(thisBank.name[i], thisBank.gid[i], thisBank.type[i], thisBank.description[i]);
			}
		}
		
	}
	dgtBank.orderNames();
	return dgtBank;
}


// Overloaded "<<" for the class 'bank'
ostream &operator<<(ostream &stream, gBank bank)
{
	cout << " >> Bank " << bank.bankName << " loaded with id " << bank.idtag << " : " << bank.bdescription << endl;
	
	for(unsigned i = 0; i<bank.name.size(); i++)
	{
		cout << "  > Variable : " << bank.name[i] << "\t id: " << bank.gid[i] << "\t type: " << bank.type[i] << " : " << bank.description[i] << endl;
	}
	cout << endl;
	return stream;
}

